### PR TITLE
[user-streams] Assign streams to gradient accum in bwd

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #167747
* #167705
* #167704
* __->__ #167513

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

File: torch/_functorch/_aot_autograd/streams.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import torch.fx
+from torch._dynamo.graph_utils import _get_flat_args
+
+from .utils import _is_backward_node_with_seq_nr, _is_forward_node_with_seq_nr
+
+
+Node = torch.fx.Node
```
Comment (Skylion007) [inline torch/_functorch/_aot_autograd/streams.py]: Mark as Node: TypeAlias =

File: torch/_functorch/_aot_autograd/streams.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+Node: TypeAlias = torch.fx.Node
+
+
+def is_gradient_acc(node: Node) -> bool:
+    return node.meta.get("is_gradient_acc", False)
+
+
+def get_device(node: Node) -> torch.device:
```
Comment (soulitzer) [inline torch/_functorch/_aot_autograd/streams.py]: not used anymore

File: torch/_functorch/_aot_autograd/streams.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+def assign_backward_streams(gm: torch.fx.GraphModule) -> None:
+    """Assigns backward streams to gradient accumulation nodes"""
+
+    for node in gm.graph.nodes:
+        if is_gradient_acc(node):
+            # Accumulation stream selection. Follow the rules from top to bottom to determine the accumulation stream:
+            # 1. Match first stream assignment encountered in the args from left to right
+            # 2. Match first stream assignment of the first user
+            gradients = _get_flat_args(node, {})
```
Comment (soulitzer) [inline torch/_functorch/_aot_autograd/streams.py:45]: Maybe worth documenting what changed from eager for specific examples

Comment (soulitzer) [reply torch/_functorch/_aot_autograd/streams.py:45]: oh if we cared about matching eager just slightly more without increasing complexity too much, we could iterate over all the nodes in reverse order and match the stream of the first user.

Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:45]: Ah so start with users, then args?

Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:45]: updated, lmk if that makes sense

Comment (soulitzer) [reply torch/_functorch/_aot_autograd/streams.py:45]: oh I don't see the latest commit haha

Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:45]: actually updated now lol lmk if that's what you mean by reverse order, I iterate through the users, then the args now.

Comment (soulitzer) [reply torch/_functorch/_aot_autograd/streams.py:45]: You'd have to also traverse through gm.graph.nodes in reverse order because otherwise if your user is an add, its annotation for stream would still be None, and you'd default to the arg's stream anyway.

Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:45]: ahh I see propagate in the other direction, fixed!

File: torch/_functorch/_aot_autograd/streams.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            # 2. Match first stream assignment encountered in the args from left to right
+            # This differs from eager in some cases:
+            # Specifically the eager code uses the autograd node to determine the stream,
+            # crucially this does not necessarily correspond to the FX graph node. For example,
+            # in the backward for an add node with a constant we will passthrough and during backward tracing,
+            # no op will be added to the FX graph, so our stream assignment will differ in this case.
+            gradients = _get_flat_args(node, {})
+            users = list(node.users.keys())
+
+            # All gradients will be on same device, they will be coerced if they were not with a .to() node
```
Comment (soulitzer) [inline torch/_functorch/_aot_autograd/streams.py:48]: Hmm .to() nodes won't have stream annotation either. Could that be problematic?

File: test/dynamo/test_streams.py [hunk 5]
```diff
@@ -524,7 +524,68 @@ def forward(self, tangents_1: "f32[2, 2]", tangents_2: "f32[2, 2]"):
...
+        self.assertExpectedInline(
+            print_graph(bw_graphs[0]),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, tangents_1: "f32[2, 2]", tangents_2: "f32[2, 2]"):
+        # Annotation: {'stream': 0}
+        mul_2: "f32[2, 2]" = torch.ops.aten.mul.Tensor(tangents_2, 2)
+
         #
+        add_2: "f32[2, 2]" = torch.ops.aten.add.Tensor(tangents_2, tangents_1);  tangents_2 = None
```
Comment (soulitzer) [inline test/dynamo/test_streams.py:583]: What stream does this one run on?

Comment (mlazos) [reply test/dynamo/test_streams.py:583]: It they don't have an annotation, the current stream of the device

File: torch/_functorch/_aot_autograd/streams.py (hunk 4 above)
Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:48]: No, they will run on the current stream of the device, and we will insert a sync if they are on a different stream from producer/consumer. Although do you think it could be a potential improvement (fewer syncs) if I try to choose a stream for them? What stream do they run on for eager?

Comment (soulitzer) [reply torch/_functorch/_aot_autograd/streams.py:48]: They always run on previous (autograd ) node in eager, so there would never be no stream sync there

Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:48]: Cool, then I will leave as is.

**Review by soulitzer — APPROVED** [review]
SGTM

but as discussed may need more handling for multi-device case since .to() are not annotated either.

File: torch/_functorch/_aot_autograd/streams.py (hunk 4 above)
Comment (mlazos) [reply torch/_functorch/_aot_autograd/streams.py:48]: Discussed offline, there was a typo in the above, I think it's possible to avoid a sync if we pick the right streams. It's possible to either have 2 syncs, 1 sync or no sync if the streams are consistent. However I need to look into how `.to()` works under the hood to sync properly.
